overlord/devicestate/firstboot_preseed_test.go

```diff
@@ -0,0 +1,211 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015-2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package devicestate_test
+
+import (
+	"os"
+	"path/filepath"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/bootloader/bootloadertest"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/overlord/devicestate"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/seed/seedtest"
+	"github.com/snapcore/snapd/testutil"
+)
+
+type firstbootPreseed16Suite struct {
+	firstBootBaseTest
+	firstBoot16BaseTest
+}
+
+var _ = Suite(&firstbootPreseed16Suite{})
+
+func checkPreseedTasks(c *C, tsAll []*state.TaskSet) {
+	// the tasks of the last taskset must be mark-preseeded, mark-seeded
+	lastTasks := tsAll[len(tsAll)-1].Tasks()
+	c.Check(lastTasks, HasLen, 2)
+	preseedTask := lastTasks[0]
+	markSeededTask := lastTasks[1]
+	c.Assert(preseedTask.Kind(), Equals, "mark-preseeded")
+	c.Check(markSeededTask.Kind(), Equals, "mark-seeded")
+
+	// mark-seeded waits for mark-preseeded
+	var waitsForPreseeded bool
+	for _, wt := range markSeededTask.WaitTasks() {
+		if wt.Kind() == "mark-preseeded" {
+			waitsForPreseeded = true
+		}
+	}
+	c.Check(waitsForPreseeded, Equals, true)
+}
+
+func checkPressedTaskStates(c *C, st *state.State) {
+	doneTasks := map[string]bool{
+		"prerequisites":        true,
+		"prepare-snap":         true,
+		"link-snap":            true,
+		"mount-snap":           true,
+		"setup-profiles":       true,
+		"update-gadget-assets": true,
```

Comment:
not for this PR but we probably need to add some checks to this one that it does nothing in preseed mode

Reply:
You mean to test that it does nothing in _seed_ mode?

Reply:
No, I mean in the task, it gets run in preseed already right now if understand this test,  but it cannot really work there in many cases, so it needs to check it  is a nop in practice when snapd is in preseed mode

Reply:
I'll look at this in a followup.